File: bms/go/go-edlib/OSADamerauLevenshteinDistance/OSADamerauLevenshteinDistance.go
package test 

import (
)

var (
)

const (
)

func OSADamerauLevenshteinDistance(str1, str2 string) int {
	// Convert string parameters to rune arrays to be compatible with non-ASCII
	runeStr1 := []rune(str1)
	runeStr2 := []rune(str2)

	// Get and store length of these strings
	runeStr1len := len(runeStr1)
	runeStr2len := len(runeStr2)
	if runeStr1len == 0 {
		return runeStr2len
	} else if runeStr2len == 0 {
		return runeStr1len
	} else if Equal(runeStr1, runeStr2) {
		return 0
	} else if runeStr1len < runeStr2len {
		return OSADamerauLevenshteinDistance(str2, str1)
	}

	// 2D Array
	row := Min(runeStr1len+1, 3)
	matrix := make([][]int, row)
	for i := 0; i < row; i++ {
		matrix[i] = make([]int, runeStr2len+1)
		matrix[i][0] = i
	}

	for j := 0; j <= runeStr2len; j++ {
		matrix[0][j] = j
	}

	var count int
	for i := 1; i <= runeStr1len; i++ {
		matrix[i%3][0] = i
		for j := 1; j <= runeStr2len; j++ {
			if runeStr1[i-1] == runeStr2[j-1] {
				count = 0
			} else {
				count = 1
			}

			matrix[i%3][j] = Min(Min(matrix[(i-1)%3][j]+1, matrix[i%3][j-1]+1),
				matrix[(i-1)%3][j-1]+count) // insertion, deletion, substitution
			if i > 1 && j > 1 && runeStr1[i-1] == runeStr2[j-2] && runeStr1[i-2] == runeStr2[j-1] {
				matrix[i%3][j] = Min(matrix[i%3][j], matrix[(i-2)%3][j-2]+1) // translation
			}
		}
	}
	return matrix[runeStr1len%3][runeStr2len]
}

func Equal(a, b []rune) bool {
	if len(a) != len(b) {
		return false
	}
	for i, v := range a {
		if v != b[i] {
			return false
		}
	}
	return true
}

func Min(a int, b int) int {
	if b < a {
		return b
	}
	return a
}

File: translations_only/success/go/go-edlib/OSADamerauLevenshteinDistance/mistral/Restart/OSADamerauLevenshteinDistance.rs

use std::cmp::min;

pub fn OSADamerauLevenshteinDistance(str1: String, str2: String) -> isize {
    let rune_str1 = str1.chars().collect::<Vec<char>>();
    let rune_str2 = str2.chars().collect::<Vec<char>>();

    let rune_str1len = rune_str1.len();
    let rune_str2len = rune_str2.len();

    if rune_str1len == 0 {
        return rune_str2len as isize;
    } else if rune_str2len == 0 {
        return rune_str1len as isize;
    } else if rune_str1 == rune_str2 {
        return 0;
    } else if rune_str1len < rune_str2len {
        return OSADamerauLevenshteinDistance(str2, str1);
    }

    let mut matrix = vec![vec![0; rune_str2len + 1]; 3];

    for i in 0..=rune_str2len {
        matrix[0][i] = i as isize;
    }

    let mut count: isize;
    for i in 1..=rune_str1len {
        matrix[i % 3][0] = i as isize;
        for j in 1..=rune_str2len {
            if rune_str1[i - 1] == rune_str2[j - 1] {
                count = 0;
            } else {
                count = 1;
            }

            matrix[i % 3][j] = min(min(matrix[(i - 1) % 3][j] + 1, matrix[i % 3][j - 1] + 1), matrix[(i - 1) % 3][j - 1] + count);

            if i > 1 && j > 1 && rune_str1[i - 1] == rune_str2[j - 2] && rune_str1[i - 2] == rune_str2[j - 1] {
                matrix[i % 3][j] = min(matrix[i % 3][j], matrix[(i - 2) % 3][j - 2] + 1);
            }
        }
    }

    matrix[rune_str1len % 3][rune_str2len]
}

pub fn Equal(a: Vec<char>, b: Vec<char>) -> bool {
    if a.len() != b.len() {
        return false;
    }

    for i in 0..a.len() {
        if a[i] != b[i] {
            return false;
        }
    }

    true
}

pub fn Min(a: isize, b: isize) -> isize {
    if b < a {
        return b;
    }

    a
}
